File: ffgraph/src/converter_video.c
#include "ffgraph.h"
#include <assert.h>
#include <string.h>

static const enum AVPixelFormat PIX_FMT = AV_PIX_FMT_RGB24;

typedef struct {
    struct SwsContext* sws_ctx;
    AVFrame* frame;
    uint8_t* buffer;
    int32_t buffer_size;
    int width;
    int height;
    double seconds;
    double frame_duration;
} VideoConverter;


static bool video_process(FFGraphConversor* ffgraphconv, const AVCodecContext* codec_ctx, const AVFrame* av_frame) {
    if (!av_frame) return false;

    VideoConverter* obj = ffgraphconv->priv_data;
    const uint8_t* const* src_data = (const uint8_t* const*)av_frame->data;

    sws_scale(
        obj->sws_ctx,
        src_data, av_frame->linesize,
        0, codec_ctx->height,
        obj->frame->data, obj->frame->linesize
    );

    obj->seconds = calculate_seconds(ffgraphconv->av_stream, av_frame);
    obj->frame_duration = av_frame->duration;

    return false;
}

static void video_destroy(VideoConverter* conv_ctx) {
    sws_freeContext(conv_ctx->sws_ctx);
    if (conv_ctx->buffer) av_free(conv_ctx->buffer);
    if (conv_ctx->frame) av_frame_free(&conv_ctx->frame);
    av_free(conv_ctx);
}

static void video_destroy_from_conversor(FFGraphConversor* ffgraphconv) {
    if (!ffgraphconv || !ffgraphconv->priv_data) return;

    video_destroy(ffgraphconv->priv_data);
    ffgraphconv->priv_data = NULL;
}


bool videoconverter_init(const AVCodecContext* codec_ctx, FFGraphConversor* ffgraphconv) {
    assert(codec_ctx);

    int size = av_image_get_buffer_size(PIX_FMT, codec_ctx->width, codec_ctx->height, 1);
    if (size < 0) {
        printf("videoconverter_init() call to av_image_get_buffer_size() failed.\n");
        return false;
    }
    uint8_t* buffer = av_malloc(size);
    if (!buffer) {
        printf("videoconverter_init() failed to allocate the buffer.\n");
        return false;
    }

    VideoConverter* obj = av_malloc(sizeof(VideoConverter));
    obj->frame = NULL;
    obj->sws_ctx = NULL;
    obj->buffer = buffer;
    obj->buffer_size = size;
    obj->width = codec_ctx->width;
    obj->height = codec_ctx->height;
    obj->seconds = -1.0;
    obj->frame_duration = codec_ctx->framerate.den / (double)codec_ctx->framerate.num;

    if (codec_ctx->framerate.num == 0 && codec_ctx->framerate.den == 1) obj->frame_duration = 0.0;

    obj->sws_ctx = sws_getContext(
        codec_ctx->width, codec_ctx->height, codec_ctx->pix_fmt,
        codec_ctx->width, codec_ctx->height, PIX_FMT,
        SWS_BILINEAR,
        NULL, NULL, NULL
    );
    if (!obj->sws_ctx) {
        printf("video_process() call to sws_getContext() failed.\n");
        goto L_failed;
    }

    obj->frame = av_frame_alloc();
    if (!obj->frame) {
        printf("videoconverter_init() call to av_frame_alloc() failed.\n");
        goto L_failed;
    }

    av_image_fill_arrays(
        obj->frame->data, obj->frame->linesize, buffer, PIX_FMT, codec_ctx->width, codec_ctx->height, 1
    );

    ffgraphconv->priv_data = obj;
    ffgraphconv->convert_cb = video_process;
    ffgraphconv->destroy_cb = video_destroy_from_conversor;
    ffgraphconv->flush_cb = NULL;
    ffgraphconv->bufferseek_cb = NULL;
    return true;

L_failed:
    video_destroy(obj);
    return false;
}

void videoconverter_get_stream_info(FFGraphConversor* ffgraphconv, int* out_width, int* out_height) {
    VideoConverter* obj = ffgraphconv->priv_data;
    *out_width = obj->width;
    *out_height = obj->height;
}

double videoconverter_read(FFGraphConversor* ffgraphconv, double seconds, void** frame_out, int32_t* frame_size_out) {
    VideoConverter* obj = ffgraphconv->priv_data;

    // check if there no frame at all
    if (obj->seconds < 0) return -1.0;

    // check if the buffered frame is future
    if (seconds >= 0 && obj->seconds > 0 && obj->seconds > seconds) {
        return -1.0;
    }

    *frame_out = obj->buffer;
    *frame_size_out = obj->buffer_size;

    return obj->seconds;
}

File: src-desktop/ffgraph/converter_video.cs
using System;
using Engine.Externals.FFGraphInterop.Interop;

namespace Engine.Externals.FFGraphInterop;

internal unsafe class VideoConverter : IFFGraphFormatConverter {

    private const AVPixelFormat FALLBACK_PIX_FMT = AVPixelFormat.AV_PIX_FMT_RGB24;

    public static bool force_rgb24 = false;

    private SwsContext* sws_ctx;
    private int width;
    private int height;
    private double seconds;
    private double frame_duration;
    private AVFrame* frame;
    private byte* rgb24_buffer;
    private AVPixelFormat pixel_format;
    private FFGraphFrame result;


    public bool Process(AVCodecContext* codec_ctx, AVFrame* av_frame, AVStream* av_stream) {
        if (av_frame == null) {
            this.seconds = -2.0;
            return true;
        }

        if (this.sws_ctx != null) {
            FFmpeg.sws_scale(
                this.sws_ctx,
                (byte**)&av_frame->data, av_frame->linesize,
                0, codec_ctx->height,
                (byte**)&this.frame->data, this.frame->linesize
            );
        } else {
            this.result.width = av_frame->width;
            this.result.height = av_frame->height;
            this.result.y = av_frame->data[0];
            this.result.u = av_frame->data[1];
            this.result.v = av_frame->data[2];
            this.result.y_stride = av_frame->linesize[0];
            this.result.u_stride = av_frame->linesize[1];
            this.result.v_stride = av_frame->linesize[2];
        }

        if (av_frame->pts != FFmpeg.AV_NOPTS_VALUE) {
            this.seconds = FFGraph.CalculateSeconds(av_stream, av_frame);
        }
        this.frame_duration = av_frame->duration;

        return false;
    }

    public void Flush() { }

    public void Seek(double seconds) { }

    public void Dispose() {
        if (this.sws_ctx != null)
            FFmpeg.sws_freeContext(this.sws_ctx);
        if (this.rgb24_buffer != null)
            FFmpeg.av_free(this.rgb24_buffer);
        if (this.frame != null)
            fixed (AVFrame** ptr = &this.frame) FFmpeg.av_frame_free(ptr);
    }


    public static VideoConverter Init(AVCodecContext* codec_ctx) {
        if (codec_ctx == null) throw new ArgumentNullException(nameof(codec_ctx));

        AVPixelFormat pixel_format = codec_ctx->pix_fmt;
        byte* rgb24_buffer = null;
        AVFrame* frame = null;

        if (force_rgb24 || pixel_format != AVPixelFormat.AV_PIX_FMT_YUV420P) {
            pixel_format = VideoConverter.FALLBACK_PIX_FMT;

            int size = FFmpeg.av_image_get_buffer_size(pixel_format, codec_ctx->width, codec_ctx->height, 1);
            if (size < 0) {
                Logger.Error("VideoConverter::Init() call to av_image_get_buffer_size() failed.");
                return null;
            }
            rgb24_buffer = (byte*)FFmpeg.av_malloc((nuint)size);
            if (rgb24_buffer == null) {
                Logger.Error("VideoConverter::Init() failed to allocate the buffer.");
                return null;
            }

            frame = FFmpeg.av_frame_alloc();
            if (frame == null) {
                Logger.Error("VideoConverter::Init() call to av_frame_alloc() failed.");
                FFmpeg.av_free(rgb24_buffer);
                return null;
            }

            frame->width = codec_ctx->width;
            frame->height = codec_ctx->height;
            frame->format = (int)pixel_format;

            FFmpeg.av_image_fill_arrays(
               (byte**)&frame->data,
               frame->linesize, rgb24_buffer,
               pixel_format,
               codec_ctx->width, codec_ctx->height,
               1
            );
        }

        VideoConverter converter = new VideoConverter() {
            sws_ctx = null,
            width = codec_ctx->width,
            height = codec_ctx->height,
            seconds = -1.0,
            frame_duration = codec_ctx->framerate.den / (double)codec_ctx->framerate.num,
            frame = frame,
            rgb24_buffer = rgb24_buffer,
            pixel_format = pixel_format,
            result = new FFGraphFrame()
        };

        if (codec_ctx->framerate.num == 0 && codec_ctx->framerate.den == 1) {
            converter.frame_duration = 0.0;
        }

        converter.result.width = codec_ctx->width;
        converter.result.height = codec_ctx->height;

        if (pixel_format == VideoConverter.FALLBACK_PIX_FMT) {
            converter.sws_ctx = FFmpeg.sws_getContext(
                codec_ctx->width, codec_ctx->height, codec_ctx->pix_fmt,
                codec_ctx->width, codec_ctx->height, pixel_format,
                0x00,
                null, null, null
            );
            if (converter.sws_ctx == null) {
                Logger.Error("VideoConverter::Init() call to sws_getContext() failed.");
                converter.Dispose();
                return null;
            }

            converter.result.rgb24 = (nint)rgb24_buffer;
        }

        return converter;
    }

    public void GetStreamInfo(ref FFGraphInfo info) {
        info.video_width = this.width;
        info.video_height = this.height;
        info.video_is_yuv420p_or_rgb24 = this.pixel_format != VideoConverter.FALLBACK_PIX_FMT;
    }

    public double Read(out FFGraphFrame frame) {
        // check if there no frame at all
        if (this.seconds < 0.0)
            frame = default;
        else
            frame = this.result;

        return this.seconds;
    }

}
